### ISSUE TITLE
systemd does not properly kill "bad" unit generators

### ISSUE THREAD

**Issue by evverx** [systemd/systemd#3380]
Standard killing procedure:
- `systemd` sends `TERM`, `CONT`, `KILL` to all processes in the control group

This is not true for generators.

For example `fork-and-hog-cpu-generator`:

``` sh
mkdir -p /etc/systemd/system-generators/

cat <<'EOF' >/etc/systemd/system-generators/fork-and-hog-cpu-generator
#!/bin/sh

sh -c 'dd if=/dev/zero of=/dev/null' &
EOF

chmod 0755 /etc/systemd/system-generators/fork-and-hog-cpu-generator

systemctl daemon-reload

# journalctl -b | grep hog-cpu
# May 28 22:29:26 systemd-testsuite systemd[44]: Spawned /etc/systemd/system-generators/fork-and-hog-cpu-generator as 45.
# May 28 22:29:26 systemd-testsuite systemd[44]: /etc/systemd/system-generators/fork-and-hog-cpu-generator succeeded.
#
# -bash-4.3# ps aufx
# USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
# root         1  0.5  0.2  38168  6124 ?        Ss   22:29   0:00 /lib/systemd/systemd systemd.unit=multi-user.target
# ...
# root        56  0.0  0.0   4508   716 ?        S    22:29   0:00 sh -c dd if=/dev/zero of=/dev/null
# root        57 95.2  0.0   4420   716 ?        R    22:29   0:04  \_ dd if=/dev/zero of=/dev/null
```

See also:
- [stop-generator](https://gist.github.com/evverx/0f07e4ebe7e98a3506a34ac723ad4b5a#stop-generator)
- [fork-and-stop-generator](https://gist.github.com/evverx/0f07e4ebe7e98a3506a34ac723ad4b5a#fork-and-stop-generator)

**Comment by poettering** [comment]
You are suggesting we should setup a cgroup for the generators and kill it afterwards? Yeah, that certainly makes sense.

**Comment by evverx** [comment]
Yeah.
Also, I think we should rename a generator executor to something like `(sd-gen)`. This will simplify debugging.
For example, I use [`offwaketime`](http://www.brendangregg.com/blog/images/2016/offwaketime-legend.png) sometimes.

[`systemctl daemon-reload (Connection timed out)`](https://github.com/systemd/systemd/issues/3353#issuecomment-222283205)

![screen shot 2016-05-31 at 05 19 05](https://cloud.githubusercontent.com/assets/10195800/15661518/d12e353a-26f0-11e6-8549-82c3b73bcca2.png)

**Comment by poettering** [comment]
Yeah, makes a ton of sense, both placing this in a cgroup, and renaming the intermediary process

**Comment by keszybz** [comment]
Do we want this? I think we can assume generators are "trusted" and shouldn't do such things. Generators have maximum privileges, and shouldn't be doing crazy things. Setting up sandboxing for containers is going to slow down and complicate things in a very hot path, without much real benefit. We already kill generators after a timeout, and this should be enough to catch most errors. "Rogue" containers have not been a problem so far.
